File: python/modules/residual/reverb_model.py
"""ReverbModel — port of AEC3 reverb_model.cc.

Mirrors docs/aec3_extracts/src/aec3/reverb_model.{cc,h}. Tiny exponential
IIR accumulator over a per-bin power spectrum:

    reverb[k] = (reverb[k] + power_spectrum[k] * scaling) * decay

Two update flavours match AEC3 source:
  * ``update_no_freq_shaping`` — scalar ``scaling`` (used in nonlinear
    mode, where scaling = echo_path_gain).
  * ``update`` — per-bin ``scaling`` (linear mode, scaling = filter
    frequency response squared).

The accumulator is shared across modes; only the update rule differs.
"""
from __future__ import annotations

import numpy as np


class ReverbModel:
    def __init__(self, n_bins: int = 257) -> None:
        self._n_bins = int(n_bins)
        self._reverb = np.zeros(self._n_bins, dtype=np.float32)

    def reset(self) -> None:
        self._reverb.fill(0.0)

    @property
    def reverb(self) -> np.ndarray:
        return self._reverb

    def update_no_freq_shaping(
        self, power_spectrum: np.ndarray, scaling: float, decay: float
    ) -> None:
        """Mirrors UpdateReverbNoFreqShaping (cc:28-39)."""
        if decay <= 0.0:
            return
        np.add(self._reverb, power_spectrum * float(scaling), out=self._reverb)
        self._reverb *= float(decay)

    def update(
        self,
        power_spectrum: np.ndarray,
        scaling: np.ndarray,
        decay: float,
    ) -> None:
        """Mirrors UpdateReverb (cc:41-52). Per-bin scaling."""
        if decay <= 0.0:
            return
        np.add(self._reverb, power_spectrum * scaling, out=self._reverb)
        self._reverb *= float(decay)

File: c_impl/src/reverb_model.c
/* reverb_model.c — C port of reverb_model.py. All-float32 arithmetic to match
 * the numpy<2 reference (float32 power_spectrum + scalar/float32 scaling →
 * three separate float32 roundings per bin). */
#include "reverb_model.h"

#include <string.h>

void reverb_model_init(ReverbModel *m, float *reverb_storage, int n_bins) {
    m->n_bins = n_bins;
    m->reverb = reverb_storage;
    memset(m->reverb, 0, (size_t)n_bins * sizeof(float));
}

void reverb_model_reset(ReverbModel *m) {
    memset(m->reverb, 0, (size_t)m->n_bins * sizeof(float));
}

void reverb_model_update_no_freq_shaping(ReverbModel *m,
                                         const float *power_spectrum,
                                         float scaling, float decay) {
    int k;
    if (decay <= 0.0f) {
        return;
    }
    for (k = 0; k < m->n_bins; ++k) {
        /* numpy: tmp = ps*scaling (f32); reverb += tmp (f32); reverb *= decay (f32) */
        float tmp = power_spectrum[k] * scaling;
        m->reverb[k] = m->reverb[k] + tmp;
        m->reverb[k] = m->reverb[k] * decay;
    }
}

void reverb_model_update(ReverbModel *m, const float *power_spectrum,
                         const float *scaling, float decay) {
    int k;
    if (decay <= 0.0f) {
        return;
    }
    for (k = 0; k < m->n_bins; ++k) {
        float tmp = power_spectrum[k] * scaling[k];
        m->reverb[k] = m->reverb[k] + tmp;
        m->reverb[k] = m->reverb[k] * decay;
    }
}
